### ISSUE TITLE
Creating a nested loop for running NIRCAM stage-3 data reduction for a large group of calibrated files

### ISSUE THREAD

**Issue by xnaxe29** [spacetelescope/jwst#7906]
Dear Team,

I am facing a specific issue while running Stage 3 of the JWST-NIRCAM calibration pipeline in a loop. 
My idea was to run the stage-3 pipeline in an iterative loop, with the first batch of reduced mosaics - "_resamplestep.fits" (while running resample step) or "_i2d.fits" (while running the whole pipeline) to act as the input for the second iteration. 

But while doing this, I am running into a specific key error for WCS 
  image3_run = image3.run(asn_file)
  File "stpipe/step.py", line 478, in run
    step_result = self.process(*args)
  "jwst/pipeline/calwebb_image3.py", line 89, in process
    input_models = self.tweakreg(input_models)
  "stpipe/step.py", line 478, in run
    step_result = self.process(*args)
  "jwst/tweakreg/tweakreg_step.py", line 284, in process
    wcsimlist = list(map(self._imodel2wcsim, g))
  "jwst/tweakreg/tweakreg_step.py", line 534, in _imodel2wcsim
    wcsinfo={'roll_ref': refang['roll_ref'],
  KeyError: 'roll_ref'

In short, what I understand is that the mosaiced image produced by resample (stage-3) cannot be feeded as a input to the stage-3 pipeline. Alternatively, I can try to stop running the pipeline at outlier_detection stage-3, but that produces individual (_crf.fits) 1-to-1 files to each stage-2 calibrated file (_cal.fits). Hence, it beats my purpose, where I would like to run the mosaicing stage in nested loop to reduce memory usage. 

Can you please help with this problem or suggest a workaround?

Thanks
Adarsh

**Comment by xnaxe29** [comment]
UPDATE:
Dear All,

While debugging, I found out that the issue is with how many WCS elements are present in the *_cal.fits generated from Stage-2 and Stage-3.

STAGE-2 *_cal.fits FILE WCS INFORMATION:
{'a_0_2': -1.5441462132232321e-06, 'a_0_3': 1.1576170066143197e-11, 'a_1_1': -1.1658681878817307e-05, 'a_1_2': 1.8007031064200423e-09, 'a_2_0': 1.8757832139863774e-06, 'a_2_1': -3.417826952546471e-11, 'a_3_0': 1.657415817888524e-09, 'a_order': 3, 'ap_0_1': 2.405415502450592e-07, 'ap_0_2': 1.537278178663311e-06, 'ap_0_3': 2.6749655825397178e-11, 'ap_1_0': -1.0120941669768866e-05, 'ap_1_1': 1.1538368508477909e-05, 'ap_1_2': -1.5820127106615878e-09, 'ap_2_0': -1.8581589719935154e-06, 'ap_2_1': -8.776317134503305e-11, 'ap_3_0': -1.6887066593033523e-09, 'ap_order': 3, 'b_0_2': -6.756973535699579e-06, 'b_0_3': 1.7274577115333498e-09, 'b_1_1': 3.5861385156452086e-06, 'b_1_2': -5.975313145423626e-11, 'b_2_0': 4.983259366236231e-06, 'b_2_1': 1.6903801198931637e-09, 'b_3_0': -3.457749648712744e-12, 'b_order': 3, 'bp_0_1': -9.737463051092451e-06, 'bp_0_2': 6.692510295532183e-06, 'bp_0_3': -1.623297819632931e-09, 'bp_1_0': 6.064879566375876e-07, 'bp_1_1': -3.5485995771322827e-06, 'bp_1_2': -6.962718679132069e-11, 'bp_2_0': -4.962750452104246e-06, 'bp_2_1': -1.830514294892991e-09, 'bp_3_0': 3.8968966246932e-11, 'bp_order': 3, 'cd1_1': 6.7901316335115e-06, 'cd1_2': 1.6120522766686e-05, 'cd2_1': 1.6053424968831e-05, 'cd2_2': -6.8489524942272e-06, 'crpix1': 1024.5, 'crpix2': 1024.5, 'crval1': 150.05766077568, 'crval2': 2.5716262633064, 'ctype1': 'RA---TAN-SIP', 'ctype2': 'DEC--TAN-SIP', 'cunit1': 'deg', 'cunit2': 'deg', 'dec_ref': 2.571627527373288, 'latpole': 2.5716262633064, 'lonpole': 180.0, 'mjdref': 0.0, 'ra_ref': 150.0576643568557, 'radesys': 'ICRS', 'roll_ref': 113.0718262130282, 's_region': 'POLYGON ICRS  150.034053889 2.561966200 150.067265914 2.548333819 150.081190247 2.580887008 150.048067449 2.595409843', 'sipiverr': 0.07087210669785854, 'sipmxerr': 0.09659505452660364, 'v2_ref': 85.939298, 'v3_ref': -493.510653, 'v3yangle': -0.054893, 'velosys': 29144.19, 'vparity': -1, 'wcsaxes': 2, 'wcsname': 'world'}


STAGE-3 *_cal.fits FILE WCS INFORMATION:
{'cdelt1': 5.555555555555556e-06, 'cdelt2': 5.555555555555556e-06, 'crpix1': 4862.306098525623, 'crpix2': 5435.775051717339, 'crval1': 150.0541440268036, 'crval2': 2.565912534533652, 'ctype1': 'RA---TAN', 'ctype2': 'DEC--TAN', 'cunit1': 'deg', 'cunit2': 'deg', 'pc1_1': -1.0, 'pc1_2': 0.0, 'pc2_1': 0.0, 'pc2_2': 1.0, 's_region': 'POLYGON ICRS  150.081177748 2.535719061 150.081179025 2.596107937 150.027106872 2.596107937 150.027108149 2.535719061', 'velosys': 29144.19, 'wcsaxes': 2}

If this problem cannot be addressed directly, is there any possible alternative solution?

Thanks in advance for answering.
Cheers,
Adarsh

**Comment by hbushouse** [comment]
Not sure exactly what you're trying to accomplish here. You're correct that images that have already been resampled can't be fed back in to do more resampling. You wouldn't want to do that anyway - never resample data more than once. Is it because you've got a very large mosaic with too many individual input images to feed in all at one time? Off the top of my head, the only way to do such a thing would be to run the resample step multiple times, each time using only a subset of either "cal" or "crf" images, but specify your own custom params for the output (resampled) image, such as the total dimensions, pixel scale, central RA/Dec, etc., which are appropriate for the final, very large mosaic that you want in the end. You'd basically be creating a number of mosaics (one on each loop) where each one is the same total size as you want for the final output, but each one would only have some portion of that giant image populated (from the subset of inputs). Once you've created multiple, partially-populated resampled images, then just combine all those resampled images using some basic image combination routine to form a single fully-populated mosaic. In order to do the simple combination of the multiple mosaics, they all have to be the same dimensions, pixel scale, etc.

**Comment by xnaxe29** [comment]
Hi,

You are right about why I am trying to do this. I basically have a large number of stage-2 calibrated files that need to be fed into the mosaicing stage.

May I know why you said, "Never resample data more than once"? The routine just takes in multiple images and produces a single undistorted image, which should make it fine to do it iteratively.

Also, can you suggest an alternative basic image combination routine to form a single fully-populated mosaic?

**Comment by melanieclarke** [comment]
Closing: please file a JWST helpdesk ticket for user support.
